src/libstd/collections/hash/table.rs

```diff
@@ -663,106 +487,98 @@ impl<K, V> RawTable<K, V> {
         self.size
     }
 
-    fn raw_buckets(&self) -> RawBuckets<K, V> {
-        RawBuckets {
-            raw: self.first_bucket_raw(),
-            hashes_end: unsafe {
-                self.hashes.offset(self.capacity as isize)
-            },
-            marker: marker::PhantomData,
-        }
+    /// Pointer to one-past-the-last key-value pair.
+    pub fn as_mut_ptr(&self) -> *mut (K, V) {
+        unsafe { self.middle.get() as *const _ as *mut _ }
     }
+}
 
-    pub fn iter(&self) -> Iter<K, V> {
-        Iter {
-            iter: self.raw_buckets(),
-            elems_left: self.size(),
-        }
-    }
+/// Rounds up to a multiple of a power of two. Returns the closest multiple
+/// of `target_alignment` that is higher or equal to `unrounded`.
+///
+/// # Panics
+///
+/// Panics if `target_alignment` is not a power of two.
+fn round_up_to_next(unrounded: usize, target_alignment: usize) -> usize {
+    assert!(target_alignment.is_power_of_two());
+    (unrounded + target_alignment - 1) & !(target_alignment - 1)
+}
 
-    pub fn iter_mut(&mut self) -> IterMut<K, V> {
-        IterMut {
-            iter: self.raw_buckets(),
-            elems_left: self.size(),
-        }
-    }
+#[test]
+fn test_rounding() {
+    assert_eq!(round_up_to_next(0, 4), 0);
+    assert_eq!(round_up_to_next(1, 4), 4);
+    assert_eq!(round_up_to_next(2, 4), 4);
+    assert_eq!(round_up_to_next(3, 4), 4);
+    assert_eq!(round_up_to_next(4, 4), 4);
+    assert_eq!(round_up_to_next(5, 4), 8);
+    assert_eq!(round_up_to_next(5, 8), 8);
+}
 
-    pub fn into_iter(self) -> IntoIter<K, V> {
-        let RawBuckets { raw, hashes_end, .. } = self.raw_buckets();
-        // Replace the marker regardless of lifetime bounds on parameters.
-        IntoIter {
-            iter: RawBuckets {
-                raw: raw,
-                hashes_end: hashes_end,
-                marker: marker::PhantomData,
-            },
-            table: self,
-        }
-    }
+#[inline]
+fn size_generic<K, V>(capacity: usize) -> usize {
+    let hash_align = min_align_of::<Option<SafeHash>>();
+    round_up_to_next(size_of::<(K, V)>() * capacity, hash_align) + size_of::<SafeHash>() * capacity
+}
 
-    pub fn drain(&mut self) -> Drain<K, V> {
-        let RawBuckets { raw, hashes_end, .. } = self.raw_buckets();
-        // Replace the marker regardless of lifetime bounds on parameters.
-        Drain {
-            iter: RawBuckets {
-                raw: raw,
-                hashes_end: hashes_end,
-                marker: marker::PhantomData,
-            },
-            table: self,
-        }
+fn checked_size_generic<K, V>(capacity: usize) -> usize {
+    let size = size_generic::<K, V>(capacity);
+    let elem_size = size_of::<(K, V)>() + size_of::<SafeHash>();
+    assert!(size >= capacity.checked_mul(elem_size).expect("capacity overflow"),
```

Comment:
I'm confused why this assert instead of just constructing `size` using checked ops?